src/dmd/statementsem.d

```diff
@@ -201,6 +202,19 @@ private extern (C++) final class StatementSemanticVisitor : Visitor
             s.exp = checkGC(sc, s.exp);
             if (s.exp.op == TOK.error)
                 return setError();
+
+            if (doCheckUnused())
```

Comment:
So you are marking the variable as referenced only if it was assigned? What about imported variables that have been passed as arguments to a function?

Reply:
> Could you explain a bit what this is supposed to do, and when the warning is supposed to trigger? @nordlow

See the tests to see what cases are currently covered.

Reply:
I'm checking three kinds of referencing; by name, by value read, by value written.